File: playwright_connection_helper.js
/**
 * Playwright Connection Helper - Brute force connection with retries
 */

const http = require('http');

/**
 * Wait for server to be available with retries
 * @param {string} url - The URL to connect to
 * @param {number} maxAttempts - Maximum number of connection attempts
 * @param {number} delay - Delay between attempts in milliseconds
 * @returns {Promise<boolean>} True if connection successful
 */
async function waitForServer(url, maxAttempts = 30, delay = 1000) {
  console.log(`Attempting to connect to ${url}...`);

  const urlObj = new URL(url);

  for (let attempt = 1; attempt <= maxAttempts; attempt++) {
    try {
      await new Promise((resolve, reject) => {
        const req = http.get({
          hostname: urlObj.hostname,
          port: urlObj.port,
          path: urlObj.pathname,
          timeout: 5000
        }, (res) => {
          if (res.statusCode === 200 || res.statusCode === 304) {
            resolve(true);
          } else {
            reject(new Error(`Status ${res.statusCode}`));
          }
        });

        req.on('error', reject);
        req.on('timeout', () => {
          req.destroy();
          reject(new Error('Request timeout'));
        });
      });

      console.log(`✓ Connected successfully on attempt ${attempt}`);
      return true;

    } catch (error) {
      if (attempt < maxAttempts) {
        console.log(`Attempt ${attempt}/${maxAttempts} failed: ${error.message.substring(0, 50)}... Retrying in ${delay/1000}s`);
        await new Promise(resolve => setTimeout(resolve, delay));
      } else {
        console.log(`✗ All ${maxAttempts} connection attempts failed`);
        return false;
      }
    }
  }

  return false;
}

/**
 * Navigate to URL with retry logic
 * @param {import('playwright').Page} page - Playwright page object
 * @param {string} url - The URL to navigate to
 * @param {number} maxAttempts - Maximum number of attempts
 * @param {number} timeout - Timeout for each attempt in milliseconds
 * @returns {Promise<boolean>} True if navigation successful
 */
async function gotoWithRetry(page, url, maxAttempts = 5, timeout = 30000) {
  for (let attempt = 1; attempt <= maxAttempts; attempt++) {
    try {
      console.log(`Navigation attempt ${attempt}/${maxAttempts} to ${url}`);
      await page.goto(url, { waitUntil: 'networkidle', timeout });
      console.log(`✓ Navigation successful on attempt ${attempt}`);
      return true;
    } catch (error) {
      if (attempt < maxAttempts) {
        console.log(`Navigation attempt ${attempt} failed: ${error.message.substring(0, 100)}... Retrying`);
        await new Promise(resolve => setTimeout(resolve, 2000));
      } else {
        console.log(`✗ All ${maxAttempts} navigation attempts failed`);
        return false;
      }
    }
  }

  return false;
}

/**
 * Ensure element is visible with retry logic
 * @param {import('playwright').Page} page - Playwright page object
 * @param {string} selector - CSS selector
 * @param {number} timeout - Timeout for each attempt in milliseconds
 * @param {number} retries - Number of retries
 * @returns {Promise<boolean>} True if element becomes visible
 */
async function ensureElementVisible(page, selector, timeout = 10000, retries = 3) {
  for (let attempt = 1; attempt <= retries; attempt++) {
    try {
      await page.locator(selector).waitFor({ state: 'visible', timeout });
      return true;
    } catch (error) {
      if (attempt < retries) {
        console.log(`Element '${selector}' not visible (attempt ${attempt}/${retries}), retrying...`);
        await new Promise(resolve => setTimeout(resolve, 1000));
      } else {
        console.log(`✗ Element '${selector}' failed to appear after ${retries} attempts`);
        return false;
      }
    }
  }

  return false;
}

/**
 * Click element with retry logic
 * @param {import('playwright').Page} page - Playwright page object
 * @param {string} selector - CSS selector
 * @param {number} maxAttempts - Maximum number of click attempts
 * @param {number} delay - Delay between attempts in milliseconds
 * @returns {Promise<boolean>} True if click successful
 */
async function clickWithRetry(page, selector, maxAttempts = 3, delay = 500) {
  for (let attempt = 1; attempt <= maxAttempts; attempt++) {
    try {
      await page.locator(selector).click({ timeout: 5000 });
      console.log(`✓ Clicked '${selector}' on attempt ${attempt}`);
      return true;
    } catch (error) {
      if (attempt < maxAttempts) {
        console.log(`Click attempt ${attempt} failed: ${error.message.substring(0, 50)}... Retrying`);
        await new Promise(resolve => setTimeout(resolve, delay));
      } else {
        console.log(`✗ Failed to click '${selector}' after ${maxAttempts} attempts`);
        return false;
      }
    }
  }

  return false;
}

/**
 * Wait for network to be idle with retry logic
 * @param {import('playwright').Page} page - Playwright page object
 * @param {number} timeout - Timeout in milliseconds
 * @param {number} maxAttempts - Maximum number of attempts
 * @returns {Promise<boolean>} True if network became idle
 */
async function waitForNetworkIdle(page, timeout = 30000, maxAttempts = 3) {
  for (let attempt = 1; attempt <= maxAttempts; attempt++) {
    try {
      await page.waitForLoadState('networkidle', { timeout });
      console.log(`✓ Network idle on attempt ${attempt}`);
      return true;
    } catch (error) {
      if (attempt < maxAttempts) {
        console.log(`Network idle wait failed (attempt ${attempt}): ${error.message.substring(0, 50)}...`);
        await new Promise(resolve => setTimeout(resolve, 1000));
      } else {
        console.log(`✗ Network did not become idle after ${maxAttempts} attempts`);
        return false;
      }
    }
  }

  return false;
}

module.exports = {
  waitForServer,
  gotoWithRetry,
  ensureElementVisible,
  clickWithRetry,
  waitForNetworkIdle
};

File: playwright_connection_helper.py
#!/usr/bin/env python3
# -*- coding: utf-8 -*-
"""
Playwright Connection Helper - Brute force connection with retries
"""

import time
import sys
import requests
from playwright.sync_api import Page, Browser, BrowserContext

def wait_for_server(url: str, max_attempts: int = 30, delay: float = 1.0) -> bool:
    """
    Brute force connection to server with retries

    Args:
        url: The URL to connect to
        max_attempts: Maximum number of connection attempts
        delay: Delay between attempts in seconds

    Returns:
        True if connection successful, False otherwise
    """
    print(f"Attempting to connect to {url}...")

    for attempt in range(1, max_attempts + 1):
        try:
            response = requests.get(url, timeout=5)
            if response.status_code == 200:
                print(f"✓ Connected successfully on attempt {attempt}")
                return True
        except requests.exceptions.RequestException as e:
            if attempt < max_attempts:
                print(f"Attempt {attempt}/{max_attempts} failed: {str(e)[:50]}... Retrying in {delay}s")
                time.sleep(delay)
            else:
                print(f"✗ All {max_attempts} connection attempts failed")
                return False

    return False


def goto_with_retry(page: Page, url: str, max_attempts: int = 5, timeout: int = 30000) -> bool:
    """
    Navigate to URL with retry logic

    Args:
        page: Playwright page object
        url: The URL to navigate to
        max_attempts: Maximum number of attempts
        timeout: Timeout for each attempt in milliseconds

    Returns:
        True if navigation successful, False otherwise
    """
    for attempt in range(1, max_attempts + 1):
        try:
            print(f"Navigation attempt {attempt}/{max_attempts} to {url}")
            page.goto(url, wait_until='networkidle', timeout=timeout)
            print(f"✓ Navigation successful on attempt {attempt}")
            return True
        except Exception as e:
            if attempt < max_attempts:
                print(f"Navigation attempt {attempt} failed: {str(e)[:100]}... Retrying")
                time.sleep(2)
            else:
                print(f"✗ All {max_attempts} navigation attempts failed")
                return False

    return False


def ensure_element_visible(page: Page, selector: str, timeout: int = 10000, retries: int = 3) -> bool:
    """
    Ensure element is visible with retry logic

    Args:
        page: Playwright page object
        selector: CSS selector or locator string
        timeout: Timeout for each attempt in milliseconds
        retries: Number of retries

    Returns:
        True if element becomes visible, False otherwise
    """
    for attempt in range(1, retries + 1):
        try:
            element = page.locator(selector)
            element.wait_for(state='visible', timeout=timeout)
            return True
        except Exception as e:
            if attempt < retries:
                print(f"Element '{selector}' not visible (attempt {attempt}/{retries}), retrying...")
                time.sleep(1)
            else:
                print(f"✗ Element '{selector}' failed to appear after {retries} attempts")
                return False

    return False


def click_with_retry(page: Page, selector: str, max_attempts: int = 3, delay: float = 0.5) -> bool:
    """
    Click element with retry logic

    Args:
        page: Playwright page object
        selector: CSS selector or locator string
        max_attempts: Maximum number of click attempts
        delay: Delay between attempts

    Returns:
        True if click successful, False otherwise
    """
    for attempt in range(1, max_attempts + 1):
        try:
            element = page.locator(selector)
            element.click(timeout=5000)
            print(f"✓ Clicked '{selector}' on attempt {attempt}")
            return True
        except Exception as e:
            if attempt < max_attempts:
                print(f"Click attempt {attempt} failed: {str(e)[:50]}... Retrying")
                time.sleep(delay)
            else:
                print(f"✗ Failed to click '{selector}' after {max_attempts} attempts")
                return False

    return False


def wait_for_network_idle(page: Page, timeout: int = 30000, max_attempts: int = 3) -> bool:
    """
    Wait for network to be idle with retry logic

    Args:
        page: Playwright page object
        timeout: Timeout in milliseconds
        max_attempts: Maximum number of attempts

    Returns:
        True if network became idle, False otherwise
    """
    for attempt in range(1, max_attempts + 1):
        try:
            page.wait_for_load_state('networkidle', timeout=timeout)
            print(f"✓ Network idle on attempt {attempt}")
            return True
        except Exception as e:
            if attempt < max_attempts:
                print(f"Network idle wait failed (attempt {attempt}): {str(e)[:50]}...")
                time.sleep(1)
            else:
                print(f"✗ Network did not become idle after {max_attempts} attempts")
                return False

    return False
